ThreadName Enricher added for Net45 and Standard2.0

StefanOssendorf opened the pull request:
ThreadName Enricher added for Net45 and NetStandard2.0.
Unfortunately Thread is not available in NetStandard1.0.

Changed the copyright statement to include 2019.

Question before merge:
At the moment in the output of a non set thread name will be `ThreadName=null`. I could think of an overload for the `EnrichWithThreadName(string nameIfNotSet)` method to specify a default value instead of `null`. Thoughts?

StefanOssendorf committed:
ThreadName Enricher added for Net45 and Standar2.0

ThreadName Enricher added for Net45 and NetStandard2.0.
Unfortunately Thread is not available in NetStandard1.0.

Changed the copyright statement to include 2019.

tsimbalar commented:
Nice !

There might be some drawbacks to this approach, but maybe another option when `Thread.Name` is `null` would be to fall back to the Thread's Id so that we always have something to display ?

nblumhardt commented:
Thread names are common but not ubiquitous; I think if I were using this, I'd probably prefer not to end up with a meaningless `ThreadName = null` property on events from unnamed threads. Likewise, I'd probably still record `ThreadId` anyway, since names are not necessarily unique, so duplicating the thread id in the name seems redundant.

I guess I'm viewing the `ThreadName` as a useful piece of information to inspect, but not seeing it used as much for as correlation id for retrieving events from the same thread. What do you think?

StefanOssendorf commented:
I agree it's only a useful piece of information to inspect.

I don't know who should make the choice for the "what ifs"?  

Another (more complex) approach would be to introduce some kind of settings object. Something like:  
```
ThreadNameSettings {
  bool IncludeIfNameIsNull;  
  Func<string, string> GetThreadName;
}
```
or something in this direction?

nblumhardt commented:
In Serilog there are a lot of these kinds of cases, so they're normally handled by gluing together enrichers, rather than building this into each individual enricher.

I.e. in this case, if you really wanted `ThreadName` for un-named threads, you'd create and `Enrich.With()` a `new DefaultThreadNameEnricher()`, where the enricher would inspect the event and add the property if its not already present.

StefanOssendorf commented:
Mhm... okay.

So the conclusion is? Add as is or some tweaks?
I personally would not mind a `ThreadName=null` in my log since it represents the actual "state".

tsimbalar commented:
Think what @nblumhardt is to have independent, orthogonal very focused enrichers that only add properties that make sense, and avoid cluttering events with undefined properties.

`ThreadName=null` sort of means "there is no thread name" so it may as well not have property at all.

To sum up, for consistency with how Enrichers are usually designed in the Serilog ecosystem : 

- `if (ThreadName!==null)` : add a `ThreadName` with that value
- `else` : do nothing

People who want a *default* thread name when missing can do : 
```
// ... new LoggerConfiguration()...
.Enrich.WithThreadName()
// add a default value when none had been provided
.Enrich.WithProperty("ThreadName", "<no-thread-name>")
// ...
```

StefanOssendorf committed:
Only add thread name if not null

Test to only add the porperty iff Thread.Name is not null.

StefanOssendorf commented:
Thank you for your explanaition. I've modified the enricher accordingly.

nblumhardt commented on src/Serilog.Enrichers.Thread/Enrichers/ThreadNameEnricher.cs:
```diff
@@ -0,0 +1,49 @@
+using Serilog.Core;
+using Serilog.Events;
+
+namespace Serilog.Enrichers {
```
Nit - missing newline

StefanOssendorf replied:
Fixed with https://github.com/serilog/serilog-enrichers-thread/pull/11/commits/3440e6544939c15da2c926fdbfb3038624a22aa8

nblumhardt commented on src/Serilog.Enrichers.Thread/Enrichers/ThreadNameEnricher.cs:
```diff
@@ -0,0 +1,49 @@
+using Serilog.Events;
+
+namespace Serilog.Enrichers {
+#if NET45 || NETSTANDARD2_0
```
Typically instead of using target frameworks for conditional compilation, we've been using `<DefineProperties>` in the CSPROJ file to define a specific constant, like `THREAD_NAME`, to indicate that the feature is present. This makes it easier to correctly add and change targeted frameworks in future versions. See e.g.: https://github.com/serilog/serilog-sinks-file/blob/dev/src/Serilog.Sinks.File/Serilog.Sinks.File.csproj#L43 and https://github.com/serilog/serilog-sinks-file/blob/dev/src/Serilog.Sinks.File/Sinks/File/SharedFileSink.AtomicAppend.cs#L15

StefanOssendorf replied:
Fixed with https://github.com/serilog/serilog-enrichers-thread/pull/11/commits/55cfd64d281e73cef12601831c38669d8d18a3f2

That's nice. I learned something new :-)

nblumhardt commented on src/Serilog.Enrichers.Thread/Enrichers/ThreadNameEnricher.cs:
```diff
@@ -0,0 +1,49 @@
+    /// </summary>
+    public class ThreadNameEnricher : ILogEventEnricher 
+    {
+
```
Nit - extra newline

StefanOssendorf replied:
Fixed with https://github.com/serilog/serilog-enrichers-thread/pull/11/commits/3440e6544939c15da2c926fdbfb3038624a22aa8

nblumhardt commented on src/Serilog.Enrichers.Thread/Enrichers/ThreadNameEnricher.cs:
```diff
@@ -0,0 +1,49 @@
+        /// </summary>
+        public const string ThreadNamePropertyName = "ThreadName";
+
+
```
Nit - extra newline

StefanOssendorf replied:
Fixed with https://github.com/serilog/serilog-enrichers-thread/pull/11/commits/3440e6544939c15da2c926fdbfb3038624a22aa8

nblumhardt commented on src/Serilog.Enrichers.Thread/Enrichers/ThreadNameEnricher.cs:
```diff
@@ -0,0 +1,49 @@
+            var threadName = Thread.CurrentThread.Name;
+            if (threadName != null) 
+            {
+                logEvent.AddPropertyIfAbsent(new LogEventProperty(ThreadNamePropertyName, new ScalarValue(Thread.CurrentThread.Name)));
```
Since we've already assigned `threadName` we could use it here instead of calling `Thread.CurrentThread.Name` (which will likely require another thread-local-storage call).

StefanOssendorf replied:
Right. Fixed with https://github.com/serilog/serilog-enrichers-thread/pull/11/commits/3440e6544939c15da2c926fdbfb3038624a22aa8

nblumhardt reviewed (commented):
Looks great, just a few nitpicky comments. Cheers!

StefanOssendorf committed:
Changes to resprect code style

Extra lines removed
Curly braces position fixed (new line not same line)
Variable name used instead of property of Thread.

StefanOssendorf committed:
THREAD_NAME feature constant added for conditional compilation

Added the feature constant THREAD_NAME to Framework4.5 and Netstandard2.0

nblumhardt committed:
Target framework names -> THREAD_NAME constant

nblumhardt commented:
Thanks @StefanOssendorf !

StefanOssendorf commented:
You're welcome :-)
And damn I missed the constant within the extension method >_> - sorry for that ^^".
